source/blog/2018-05-24-podman-userns.md

```diff
@@ -0,0 +1,157 @@
+container, but no privileges if they escape the container.
+
+READMORE
+
+[User namespaces](https://lwn.net/Articles/532593/) are a kernel
+mechanism for allowing a process to see a different uids/gids mapping
+than it has in reality.  It permits mapping multiple ranges of
+uids/gids from the host to completely different values into the
+container process.
+
+This enables things like having `root` in the container but the
+process actually runs with a UID that is not root outside of the
+container. This means if the processes in the container manages to
+break out of the container, they will be treated as normal
+non-privileged processes on the host.  Of course other security
+mechanisms like SELinux would also still be in effect, but this adds
+another layer of protection.
+
+Another interesting security aspect is that each container can have
+its own uids/gids mapped from the host.  In case of breakage, a
+container not only has a more limited attack area to the host where it
+has no root access, but other containers have another level of
+security as their uids/gids are different.  Discretionary Access
+Control, user permissions,  at its best.
+
+# Changes to the Podman CLI
+
+Podman offers two ways to use user namespaces:
+1) Directly specify the mappings
+
+Podman can specify the uids/gids directly using --uidmap and --gidmap.
+
+The first way is the most immediate, to verify what happens in a
+container we check the /proc/self/uid_map file.  It shows the uid
+mappings in place for the specified process.
+
+```
+# cat /proc/self/uid_map
+         0          0 4294967295
+
+# podman  run --uidmap=0:100000:70000 --gidmap=0:200000:70000 alpine cat /proc/self/uid_map
+         0          100000      70000
+```
+
+The uid_map table has three columns, the first one is the initial ID
+for the range in the new namespace, the second one is the initial ID
+in the parent namespace, and the third is the size of the mapping.  In
+this case, we have specified only one mapping where the UIDs [0-6999]
+in the container are mapped to the host UIDs [100000-17000].
+Similarly for the GIDs, we can look up the gid_map file.
+
+
+The process has a different UID/GID inside the container than it has
+from the host:
+
+```
+# bin/podman run --rm  --uidmap=0:11111:70000 --gidmap=0:20000:70000 alpine sh -c 'id - u; sleep 10'
+0
+```
+
+(from the host)
+
+```
+$ pgrep -U 11111 -a
+27513 sh -c id -u; sleep 10
+27574 sleep 10
+```
+
+### Use /etc/subuid and /etc/subgid
+Podman can specify the mappings defined in /etc/subuid and /etc/subgid
+with the --subuidname and --subgidname options.
+On my system I have allocated 65536 UIDs/GIDs starting at 110000 for
+the user “gscrivano”.
+
+```
+# grep gscrivano /etc/subuid
+gscrivano:110000:65536
+# grep gscrivano /etc/subgid
+gscrivano:110000:65536
+```
+
+So I can use these settings as:
+```
+# podman  run --subuidname gscrivano --subgidname=gscrivano \
+  alpine cat /proc/self/uid_map
```

Comment:
nit: I'd drop this line wrapping to stay consistent with the earlier examples where you don't wrap.